libsolidity/ast/Types.cpp

```diff
@@ -1611,6 +1550,21 @@ TypeResult ContractType::unaryOperatorResult(Token _operator) const
 		return nullptr;
 }
 
+vector<Type const*> CompositeType::fullDecomposition() const
+{
+	vector<Type const*> res = {this};
+	set<string> seen = {richIdentifier()};
+	for (size_t k = 0; k < res.size(); ++k)
+		if (auto composite = dynamic_cast<CompositeType const*>(res[k]))
+			for (Type const* next: composite->decomposition())
+				if (seen.count(next->richIdentifier()) == 0)
+				{
+					seen.insert(next->richIdentifier());
+					res.push_back(next);
+				}
```

Comment:
You could use a mapping here. I remember that one of the previous iterations of #9746 actually used a set.

Reply:
> You could use a mapping here.

I'm missing the point :). What do you mean? (I could use `unsorted_set` instead of `set`, but it is something different.)

Reply:
Sorry, looks like I used Solidity terminology by mistake. I should have said `map` :) I was just suggesting to use `map<string, Type const*>`, to replace both `seen` and `res`.

The linear search here bothers me a bit so `set`/`map` would be better but in practice huge types are rare so maybe it's not worth the hassle. I suggested it mostly because the change would not be big anyway so why not.

Reply:
Sure, I thought that `mapping` is `map`. But I do not see a good way to use it.

Let's for the sake of the argument ignore linear search and be happy with **O(N^2)** overall. 

Still, while iterating over a growing vector is easy, how do we do it over a growing map? I must be missing something and quite curious to see how it can be done.

Reply:
Oh, that's what you meant. Looks like I overlooked that small detail :) Sorry.

Well, technically you could do it with a map but it would be pretty ugly and more complex than what you have now so not really worth it. The idea would be to use a special comparator to keep the already visited N elements at the start of the map (might require adding prefixes to the strings or putting them in a `tuple` with a flag) and in each iteration visit `begin() + N` and increase the `N`. Iterate until N covers the whole map. And yeah, this `begin() + N` would eat the whole gain from not having linear search. We could get around it by having a `map` and a `set` but this is getting ridiculous so let's stay with a vector :)

Reply:
Looking at the code again, I see you're using a `set` now so there's no linear search. That's perfect and better than my `map` suggestion :)